### ISSUE TITLE
Missing dependency in 2.0.0

### ISSUE THREAD

**Issue by mvanini** [CycloneDX/cyclonedx-maven-plugin#54]
There seems to be a missing dependency in latest release. Since few hours ago we seen a lot of failure on our CIs about invocation of cyclonedx-maven-plugin via cli, like that:

`[ERROR] Failed to execute goal org.cyclonedx:cyclonedx-maven-plugin:2.0.0:makeAggregateBom (default-cli) on project MyProject: Execution default-cli of goal org.cyclonedx:cyclonedx-maven-plugin:2.0.0:makeAggregateBom failed: Plugin org.cyclonedx:cyclonedx-maven-plugin:2.0.0 or one of its dependencies could not be resolved: Could not find artifact com.github.everit-org.json-schema:org.everit.json.schema:jar:1.12.1 in my-mirror  (https://nexus.example.com/nexus/repository/my-mirror) -> [Help 1]`

I searched **com.github.everit-org.json-schema:org.everit.json.schema** on search.maven.org and i didn't found it, even for past versions... is groupdId and artifactId correct?

**Comment by coderpatros** [comment]
Hi @mvanini 

That dependency is pulled in via jitpack.io

I couldn't reproduce it locally, but I don't use maven or java normally. So maybe I'm doing it wrong. But I'm sure @stevespringett will be all over this when he comes online.

**Comment by coderpatros** [comment]
Is there any interesting information about your CI environment that you can share? And do you also hit it running locally?

**Comment by mvanini** [comment]
Hi @coderpatros , maven by default resolve dependency via https://repo1.maven.org (which is the backend of search.maven.org webapp which I mentioned before).
In our CI environment we have a corporate maven repository (Sonatype Nexus) which is also used as a proxy of all major public repository, repo1.maven.org included.
For our scenario we can add a mirror for jitpack in our Nexus but is a workaround not really a solution: not anyone that use this plugin has a local maven repository to store that dependency or possibility to mirror jitpack.

I think that all dependencies needed by a maven plugin must be publicy available on the same repository where maven look for it.

> Is there any interesting information about your CI environment that you can share? And do you also hit it running locally?
Yes, I have the same result if I run it locally

**Comment by coderpatros** [comment]
Yeah, I probably don't know enough about maven to help on this one. Maven picked it up fine for me on a test project. But that was without a corporate maven repository. I know that jitpack.io is added as a repository in the pom of the core java package, which the maven plugin uses.

**Comment by stevespringett** [comment]
> I think that all dependencies needed by a maven plugin must be publicly available on the same repository where maven look for it.

Not every dependency resides in Maven Central. The JSON schema validator is a dependency of CycloneDX Java Core which has the necessary repositories entry in the POM so that resolution can occur from non-default repositories.

https://github.com/CycloneDX/cyclonedx-core-java/blob/master/pom.xml#L103

It appears your environment however, is ignoring this. This might be a Nexus proxy configuration issue, but I'm not entirely sure. I personally do not use Nexus proxy since provenance information becomes opaque.

The GitHub action that builds the Maven plugin would have failed if the dependency could not be resolved. Additionally, some of my projects, also built with GH actions or Travis, resolve the component properly.

**Comment by mvanini** [comment]
I think that the mistake resides in the mirror use in our settings: if I declare a mirror in my settings.xml then all requests to resolve a dependency will be redirected through that mirror which, in our case, doesn't contains a proxy for the gitpack repository.

It's strange that it faults only on version 2.0.0, our config it's the same since years. Is this a new declared dependency in latest release?

> I personally do not use Nexus proxy since provenance information becomes opaque.

In an enterprise scenario with many projects, each one with his dependencies, having a local maven repository for public dependencies saves us a huge outbound network traffic to internet.

**Comment by stevespringett** [comment]
> Is this a new declared dependency in latest release?

Yes. It's necessary for JSON validation. Out of only a handful of libraries available for Java, everit-org is the most widely used and the only one that relies solely on the org.json library. The others use gson or jackson databind which I try to avoid, especially for a Mavan plugin. The only other alternative was AGPL licensed, which is toxic in most enterprises.

So the everit-org library is ideal for use with the Maven plugin both from an architecture and license perspective, but just so happens to not be hosted on Maven Central.

**Comment by mvanini** [comment]
I discovered in everit-org/json-schema#360 that it seems the dependency is published even on maven-central but with a different groupId/artifactId (why????):
```
        <dependency>
            <groupId>com.github.erosb</groupId>
            <artifactId>everit-json-schema</artifactId>
            <version>1.12.1</version>
        </dependency>
```
You can verify on https://search.maven.org/artifact/com.github.erosb/everit-json-schema

Could you consider to switch to that GAV for the next releases?

**Comment by stevespringett** [comment]
That artifact is a fork of the main project. Forks can introduce unnecessary security risk since they need to be consistently updated whenever the upstream component releases a new version. Forks are also less likely to be reviewed by security researchers and are a target for potential malware.

**Comment by rajesh2887** [comment]
> There seems to be a missing dependency in latest release. Since few hours ago we seen a lot of failure on our CIs about invocation of cyclonedx-maven-plugin via cli, like that:
> 
> `[ERROR] Failed to execute goal org.cyclonedx:cyclonedx-maven-plugin:2.0.0:makeAggregateBom (default-cli) on project MyProject: Execution default-cli of goal org.cyclonedx:cyclonedx-maven-plugin:2.0.0:makeAggregateBom failed: Plugin org.cyclonedx:cyclonedx-maven-plugin:2.0.0 or one of its dependencies could not be resolved: Could not find artifact com.github.everit-org.json-schema:org.everit.json.schema:jar:1.12.1 in my-mirror (https://nexus.example.com/nexus/repository/my-mirror) -> [Help 1]`
> 
> I searched **com.github.everit-org.json-schema:org.everit.json.schema** on search.maven.org and i didn't found it, even for past versions... is groupdId and artifactId correct?

after manually uploaded this com.github.everit-org.json-schema:org.everit.json.schema:jar:1.12.1 jar to my maven repo its working fine for me.

**Comment by rajesh2887** [comment]
can anyone help on this below error.while trying to build from Jenkins pipeline getting below error.

2020-07-16 13:38:53 [ERROR] Failed to execute goal org.cyclonedx:cyclonedx-maven-plugin:2.0.0:makeAggregateBom (default-cli) on project "myproject": Execution default-cli of goal org.cyclonedx:cyclonedx-maven-plugin:2.0.0:makeAggregateBom failed: java.net.ConnectException: Connection timed out: connect -> [Help 1]

**Comment by pachulo** [comment]
We were having the same problem.

> I think that the mistake resides in the mirror use in our settings: if I declare a mirror in my settings.xml then all requests to resolve a dependency will be redirected through that mirror which, in our case, doesn't contains a proxy for the gitpack repository.

@mvanini in our case, changing the config from this:
```
....
    <mirror>
       <id>mirror</id>
      <mirrorOf>*</mirrorOf>
      <name>MyCompany Nexus Mirror</name>
      <url>https://nexus.mycompany.net/content/groups/public</url>
    </mirror>
....
```

To this:
```
....
    <mirror> 
      <id>mirror</id> 
      <mirrorOf>central</mirrorOf> 
      <name>MyCompany Nexus Mirror</name> 
      <url>https://nexus.mycompany.net/content/groups/public</url> 
    </mirror>
....
```
(Mind the `mirrorOf` change) solved the issue.

More info: https://maven.apache.org/guides/mini/guide-mirror-settings.html

**Comment by bigboban** [comment]
Same error for me in gradle build:

```
08:30:37  > Could not resolve all artifacts for configuration ':classpath'.
08:30:37     > Could not find com.github.everit-org.json-schema:org.everit.json.schema:1.12.1.
08:30:37       Searched in the following locations:
08:30:37         - https://plugins.gradle.org/m2/com/github/everit-org/json-schema/org.everit.json.schema/1.12.1/org.everit.json.schema-1.12.1.pom
08:30:37       If the artifact you are trying to retrieve can be found in the repository but without metadata in 'Maven POM' format, you need to adjust the 'metadataSources { ... }' of the repository declaration.
08:30:37       Required by:
08:30:37           project : > org.cyclonedx.bom:org.cyclonedx.bom.gradle.plugin:1.2.1 > com.cyclonedx:cyclonedx-gradle-plugin:1.2.1 > org.cyclonedx:cyclonedx-core-java:3.0.5
```

I don't understand - issue exists more than half of year and still not fixed?

**Comment by patrickp-bosch** [comment]
We ran into the same issue with our CI setup that uses a mirror in front of the public repositories.
It is unfortunate because we noticed at the very last step, when the changes were tested and pushed, and then the build suddenly failed. It would be really convenient not to have to research what causes the failure an then look for a workaround.

**Comment by stevespringett** [comment]
The most common cause is a mis-configured Nexus or Artifactory repo. The component is question does not reside in Maven Central. It resided in Jitpack.

**Comment by prabhu** [comment]
This particular issue is causing an headache for many users. For nexus, the settings suggest here works https://github.com/jitpack/jitpack.io/issues/506#issuecomment-267338823

**Comment by stevespringett** [comment]
With version 2.3.0 of the Maven plugin, this issue no longer exists.
